doc/igmpproxy.conf.5.in

```diff
@@ -161,6 +161,21 @@ You may also specify whitelist entries for the upstream interface. Only igmp mem
 for explicitly whitelisted multicast groups will be sent out on the upstream interface. This
 is useful if you want to use multicast groups only between your downstream interfaces, like SSDP
 from a UPnP server.
+
+This option can be combined with
+.B blacklist
+for fine-grained control.
+.RE
+
+.B blacklist
+.I networkaddr
+.RS
+Defines a blacklist for multicast groups. Similar to
+.B whitelist
+except that if a blacklist entry is defined, all igmp membership reports for
+that multicast group will be ignored.
+
+Note that if whitelist and blacklist overlap, blacklist will take precedence.
```

Comment:
This looks to be strange behavior. Specially if you have big blacklist and need to whitelist some particular address from it.

Why not rather to take standard policy that the last matched rule take effect? This can be more useful for different kind of configurations.

This needs to answer two additions questions:
1. What happens when no whilelist / blacklist is specified? Everything is allowed or blocked? Allowed seems to be reasonable choice as it is backward compatible.
2. What is the default policy when particular address is not specified at all? Currently it is to block when whitelist keyword is used... Now I come up with idea that default policy may be blocked if all specified policy keywords are only whitelist (and at least one is specified), otherwise default is allowed ... this is backward compatible but I'm not sure how "sane" it is.

Reply:
> This looks to be strange behavior. Specially if you have big blacklist and need to whitelist some particular address from it.

As it is, `whitelist` is currently doing two things: 1.) it ignores all igmp membership reports, 2.) except those whitelisted. `blacklist` does only one thing: it ignores blacklisted igmp membership reports. I chose to not only keep the behaviour of whitelist, but also make it independent from blacklist. Behaviour of whitelist can be changed, but we would risk breaking existing setups.

> Why not rather to take standard policy that the last matched rule take effect? This can be more useful for different kind of configurations.

If I'm not mistaken this would mean that the behaviour of `whitelist` changes depending on if there is a `blacklist` rule. See above.
Additionally, order of the rules is lost as whitelist and blacklist are stored in separate linked lists. For the 'last match rule' to work I need to either add an additional struct member (rule numer) to each list or put all rules in a single linked list and mark them (allow/block), and write additional checks. This would make the code more complex/fragile. My proposal is to maintaing the old behaviour of whitelist, and use blacklist to additionally ignore igmp membership reports. What do you think of the following change.

```suggestion
blacklist can be used to ignore additional igmp membership reports.
```

> This needs to answer two additions questions:

Guess it makes sense to add the following to the manpage.

```
Overview on the use of whitelist and blacklist:

whitelist    blacklist     igmp membership reports
not used     not used      allow all
not used     used          allow all except those blacklisted
used         not used      block all except those whitelisted
used         used          block all except those whitelisted, and not blacklisted
```

Reply:
> If I'm not mistaken this would mean that the behaviour of whitelist changes depending on if there is a blacklist rule. See above.

Yes.

> Additionally, order of the rules is lost as whitelist and blacklist are stored in separate linked lists. For the 'last match rule' to work I need to either add an additional struct member (rule numer) to each list or put all rules in a single linked list and mark them (allow/block), and write additional checks.

Seems that change could be easy. Just extend `struct SubnetList` with a new member which says that it is blacklist or whitelist. `struct SubnetList` is used only for filtering and currently only whitelist operation.

And then it is just needed to traverse whole link list once and to the end and use rule which matches last time.

Reply:
Thanks for your feedback. Really helpful! I extended `struct SubnetList` and changed to last-rule-wins.

Changes have been pushed.